src/mlpack/methods/cf/cf_main.cpp

```diff
@@ -101,8 +105,8 @@ PARAM_DOUBLE_IN("min_residue", "Residue required to terminate the factorization"
     " (lower values generally mean better fits).", "r", 1e-5);
 
 // Load/save a model.
-PARAM_MODEL_IN(CF, "input_model", "Trained CF model to load.", "m");
-PARAM_MODEL_OUT(CF, "output_model", "Output for trained CF model.", "M");
+PARAM_MODEL_IN(CF<>, "input_model", "Trained CF model to load.", "m");
```

Comment:
@zoq What do you think would be the right way to mention the DecompositionPolicy for input_model and output_model? Right now they are being set to default policy NMF even for objects with other policy, causing error.

Reply:
I was thinking we could do ssomething similar as done for the pca executable:

https://github.com/mlpack/mlpack/blob/095f784a30742ff64f4616c362475d6c8ec8ea15/src/mlpack/methods/pca/pca_main.cpp#L68-L70

What do you think?

Reply:
Maybe I am missing something, but PCA doesn't have method to load/save model, and thus doesn't have a PARAM_MODEL_IN/OUT parameter, and this is what I am concerned about in CF.

Reply:
In other cases where this has come up (it does in `LogisticRegression` for example) I separate the template parameter used for the training of the model.  So, e.g., maybe `CF` has no template parameters but `CF::Train()` does, since the decomposition is only used at training time.

If that doesn't work (and I haven't spent a long time thinking about it, so maybe it doesn't), you might try some wrapper model that can hold all `CF<...>` types that might be used.  See, e.g., `AdaBoostModel` in `src/mlpack/methods/adaboost/adaboost_model.hpp`, then you can just serialize that.

Reply:
Good point, we don't have to templatize the complete class, ``Train()`` should be sufficient, what do you think @haritha1313 @Wenhao-H

Reply:
I agree at the moment we don't need to templatize CF class. But for rating normalization that I'm working on, CF needs to be class template because the normalization object needs to be saved as a class member. I will write a class like `AdaBoostModel` to solve this.  So for now, I think templatizing only `CF::Train()` would be enough.

Reply:
Yes, I have done the necessary edits to templatize only `CF::Train()`. I'll push them soon. @Wenhao-H I have taken a look into how `AdaBoostModel` has been written, so in case your immediate work needs CF class as templatized itself, I can do that too. Just let me know.

Reply:
@haritha1313 The implementation of the model wrapper class (called `CFModel` maybe) is determined by the class template parameter (which would be `NormalizationType` in my case). So I think it would be easier for me to work on that. Thanks for offering the help though:)

Reply:
Sounds like a good solution to me.